activerecord/lib/active_record/insert_all.rb

```diff
@@ -69,13 +75,23 @@ def readonly_columns
       end
 
       def unique_by_columns
-        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+        Array(unique_index&.columns)
       end
 
-      def primary_keys
-        Array.wrap(model.primary_key)
+      def find_index_for(unique_index)
+        match = Array(unique_index).map(&:to_s).sort
+
+        if index = indexes.find { |i| match.include?(i.name) || i.columns.sort == match }
```

Comment:
Now we could verify whether or not a unique index.

```suggestion
        if index = indexes.find { |i| i.unique && (match.include?(i.name) || i.columns.sort == match) }
```

Reply:
Great point! I thought about and forgot to do it 😄